qiskit/dagcircuit/dagcircuit.py

```diff
@@ -319,6 +319,113 @@ def _add_wire(self, wire):
         else:
             raise DAGCircuitError(f"duplicate wire {wire}")
 
+    def remove_clbits(self, *clbits):
+        """
+        Remove classical bits from the circuit. Only the bits that are idle and not
+        referenced by any classical registers will be removed. Other bits are ignored.
+
+        Raises:
+            DAGCircuitError: a clbit is not a Clbit.
+
+        Returns:
+            set(Clbit): The set of Clbits that were removed.
+        """
+        if any(not isinstance(clbit, Clbit) for clbit in clbits):
+            raise DAGCircuitError("not a Clbit instance.")
+
+        # ignore clbits not in circuit
+        clbits = set(clbits).intersection(self.clbits)
+
+        # ignore clbits referenced by registers
+        register_clbits = {bit for creg in self.cregs.values() for bit in creg}
+        clbits -= register_clbits
+
+        # ignore clbits that are busy
+        busy_clbits = {bit for bit in clbits if not self._is_wire_idle(bit)}
+        clbits -= busy_clbits
+
+        for clbit in clbits:
+            self._remove_idle_wire(clbit)
+            self.clbits.remove(clbit)
+
+        return clbits
+
+    def remove_cregs(self, *cregs):
+        """
+        Remove classical registers from the circuit,
+        removing idle bits that are no longer referenced by any register
+        as a result of this operation.
+
+        Raises:
+            DAGCircuitError: a creg is not a ClassicalRegister.
+        Returns:
+            set(Clbit): the bits removed from the circuit as a result of this operation.
+        """
+        if any(not isinstance(creg, ClassicalRegister) for creg in cregs):
+            raise DAGCircuitError("not a ClassicalRegister instance.")
+
+        # Ignore any cregs not in circuit.
+        cregs = set(cregs).intersection(self.cregs.values())
```

Comment:
Similar to the previous point - it feels like it'd be a bug if anyone _did_ pass a register that wasn't in a circuit, and we should shout about it.

Reply:
I'm happy to make this an error, but I do think it could be argued that this function only needs to make sure the registers provided are not in the circuit when it completes to have succeeded. If we make this an error, a re-run of this function will be an error. Not relevant for this function, but it seems like other methods in this codebase can fail but still have mutated state, so I wonder if there's value in providing the caller with an easy way to get themselves out of a sticky situation should one arise.

Reply:
Thanks! I think it's an interesting point. To make the decision for this PR, I ended up looking at what Python's own `List` does, and found that it raises when the element is not found. Done in e59fb8ea6c4462ead5091849c276af42cd85f917.